# Librerías

> *En hombros de gigantes*

Las librerías son bibliotecas. Un compendio de muchas líneas de código. Para que el programa funcione sólo se necesitan dos: pandas y spacy. Para que funcione mejor se requiere esta lista:

In [1]:
# INSTALACIÓN DE LIBRERÍAS

# Evitar problemas en Google Colab con spaCy
!pip install -U blis
# Instalando spaCy en Google Colab
!pip install -U pip setuptools wheel
!pip install -U spacy
# Usar modelo largo en español
!python -m spacy download es_core_news_lg

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
  Using cached blis-0.9.1-cp37-cp37m-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (7.8 MB)
  Attempting uninstall: blis
    Found existing installation: blis 0.7.9
    Uninstalling blis-0.7.9:
      Successfully uninstalled blis-0.7.9
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
thinc 8.1.5 requires blis<0.8.0,>=0.7.8, but you have blis 0.9.1 which is incompatible.
Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
  Using cached blis-0.7.9-cp37-cp37m-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (10.2 MB)
  Attempting uninstall: blis
    Found existing installation: blis 0.9.1
    Uninstalling blis-0.9.1:
      S

In [2]:
# LLAMAR LIBRERÍAS

# Agregar diccionario desde Google Drive y visualizarlo como tabla interactiva
from google.colab import drive, data_table
drive.mount("/content/drive", force_remount=True)
# Convertir diccionario, en excel, a una "tabla" más veloz
import pandas as pd
# Procesamiento de Lenguaje Natural (PLN)
import spacy
# Hacer cálculos matemáticos con vectores de palabras
import numpy as np
# Evitar conflictos con celdas vacías
import sys

Mounted at /content/drive


# Diccionario de la Reforma Protestante

## Carga

In [3]:
# - Descargando el archivo Excel del diccionario desde Google Drive
# - Conversión a Data Frame de la librería pandas
diccionario =  pd.read_excel('/content/drive/MyDrive/Diccionario/Definiciones.xlsx',
                             sheet_name='Definiciones')
# Ver una parte del diccionario
diccionario.head()

Lexema  \
0  A la nobleza cristiana de la nación alemana.   
1                                     Agustino.   
2                          Anabaptista, iglesia   
3                            Anglicana, Iglesia   
4                                    Arminiano.   

                                          Etimología Fonología  También:  \
0                                                NaN       NaN  También:   
1                                                NaN       NaN       NaN   
2  (Gr. Ανα – Contra, βαπτίζω – bautizo. Contrari...       NaN  También:   
3                        (Lat. Anglicana – Inglesa);       NaN  También:   
4                                                NaN       NaN  También:   

                        Acepciones o nombre original Fecha inicio  \
0  An den christlichen Adel deutscher Nation von ...     1520/06;   
1                                                NaN          NaN   
2                                       Anabautista;  1525/01/21;   
3                                Episcopal, Iglesia;         1532   
4                                         Wesleyano;         1610   

     Inf. Extra de Fecha Inicio Fecha de término  Inf. Extra de Fecha término  \
0                           NaN              NaN                          NaN   
1                           NaN              NaN                          NaN   
2     (Primeros re-bautizados);              NaN                          NaN   
3                       y 1536;              NaN                          NaN   
4   (como movimiento definido);              NaN                          NaN   

  Abreviatura  ... Inf. Denominacional Inf. Técnica-temática  \
0         NaN  ...               Lut.;                Dctr.;   
1         NaN  ...                 NaN                   NaN   
2         NaN  ...              Anbt.;                   NaN   
3         NaN  ...               Ang.;                   NaN   
4         NaN  ...               Arm.;              Doctrina   

  Inf. Técnica-temática2         Campo semántico       Campo semántico 2  \
0                    NaN    Literatura reformada                     NaN   
1                    NaN       Órdenes católicas                     NaN   
2                    NaN     Iglesia anabaptista  Interpretación bíblica   
3                    NaN       Iglesia anglicana                     NaN   
4                    NaN  Interpretación bíblica                     NaN   

                                          Definición  \
0  Manifiesto de Lutero contra la ICAR, donde inv...   
1               Miembro de la Orden de  San Agustín.   
2  Movimiento reformado que pugnó por un bautizo ...   
3  Movimiento reformado de Inglaterra con matices...   
4  Movimiento de la reforma iniciado por Arminio ...   

                    Complementación de la definición  \
0  Señala los tres muros del romanismo a derribar...   
1                                                NaN   
2  Manifestaban un desacuerdo a hacer juramentos,...   
3  Primera iglesia protestante estatal, por orden...   
4  La doctrina es sintetizada en los Cinco puntos...   

         Situación histórica antes y durante la Ref.  \
0  Lutero da 28 propuestas y a lo largo de ellas ...   
1                                                NaN   
2  Se gesta en tres lugares principales:  Primero...   
3  "Anglicana" refería sólo a procedencia, hasta ...   
4  Despúes del Sínodo de Dort, fueron perseguidos...   

                       Relación-aportes a la Reforma  \
0  En la Dieta de Worms se demandó a Lutero retra...   
1                                                NaN   
2  Ser rebautizado era considerado traición y se ...   
3  Mantiene elementos de la ICAR, pero no la depe...   
4  Los menonitas aceptan su doctrina. En el siglo...   

                                        Bibliografía  
0  Historia Del Cristianismo III. El Mundo Modern...  
1                                                NaN  
2  “Anabaptista” Dic

## Limpieza

Eliminamos las columnas innecesarias

In [4]:
# Simplificar diccionario. 
# Eliminar las columnas: 
diccionario = diccionario.drop(columns=[
    'Fonología',
    'También:',
    'Fecha inicio',
    'Inf. Extra de Fecha Inicio',
    'Fecha de término',
    'Inf. Extra de Fecha término',
    'Abreviatura',
    'Bibliografía'])
# Ver columnas restantes
list(diccionario.columns)

['Lexema',
 'Etimología',
 'Acepciones o nombre original',
 'Inf. Geográfica',
 'Inf. Denominacional',
 'Inf. Técnica-temática',
 'Inf. Técnica-temática2',
 'Campo semántico',
 'Campo semántico 2',
 'Definición',
 'Complementación de la definición',
 'Situación histórica antes y durante la Ref.',
 'Relación-aportes a la Reforma']

In [5]:
# Eliminar columnas con abreviaturas
diccionario = diccionario.drop(columns=[
    'Etimología',
    'Inf. Geográfica',
    'Inf. Denominacional',
    'Inf. Técnica-temática',
    'Inf. Técnica-temática2'])
# Ver columnas restantes y ver celdas con NaN (Not a Number)
diccionario.isna().sum()

Lexema                                          0
Acepciones o nombre original                   28
Campo semántico                                21
Campo semántico 2                              56
Definición                                      1
Complementación de la definición               28
Situación histórica antes y durante la Ref.    29
Relación-aportes a la Reforma                  37
dtype: int64

In [6]:
# Remplazar NaN con celda vacía (para que no dé problemas luego)
diccionario = diccionario.fillna('')
# Ver diccionario final
diccionario.head()

Lexema  \
0  A la nobleza cristiana de la nación alemana.   
1                                     Agustino.   
2                          Anabaptista, iglesia   
3                            Anglicana, Iglesia   
4                                    Arminiano.   

                        Acepciones o nombre original         Campo semántico  \
0  An den christlichen Adel deutscher Nation von ...    Literatura reformada   
1                                                          Órdenes católicas   
2                                       Anabautista;     Iglesia anabaptista   
3                                Episcopal, Iglesia;       Iglesia anglicana   
4                                         Wesleyano;  Interpretación bíblica   

        Campo semántico 2                                         Definición  \
0                          Manifiesto de Lutero contra la ICAR, donde inv...   
1                                       Miembro de la Orden de  San Agustín.   
2  Interpretación bíblica  Movimiento reformado que pugnó por un bautizo ...   
3                          Movimiento reformado de Inglaterra con matices...   
4                          Movimiento de la reforma iniciado por Arminio ...   

                    Complementación de la definición  \
0  Señala los tres muros del romanismo a derribar...   
1                                                      
2  Manifestaban un desacuerdo a hacer juramentos,...   
3  Primera iglesia protestante estatal, por orden...   
4  La doctrina es sintetizada en los Cinco puntos...   

         Situación histórica antes y durante la Ref.  \
0  Lutero da 28 propuestas y a lo largo de ellas ...   
1                                                      
2  Se gesta en tres lugares principales:  Primero...   
3  "Anglicana" refería sólo a procedencia, hasta ...   
4  Despúes del Sínodo de Dort, fueron perseguidos...   

                       Relación-aportes a la Reforma  
0  En la Dieta de Worms se demandó a Lutero retra...  
1                                                     
2  Ser rebautizado era considerado traición y se ...  
3  Mantiene elementos de la ICAR, pero no la depe...  
4  Los menonitas aceptan su doctrina. En el siglo...

## Preparar para PLN

In [7]:
# Ver columnas
diccionario.columns

Index(['Lexema', 'Acepciones o nombre original', 'Campo semántico',
       'Campo semántico 2', 'Definición', 'Complementación de la definición',
       'Situación histórica antes y durante la Ref.',
       'Relación-aportes a la Reforma'],
      dtype='object')

In [8]:
# Nuevo diccionarioPLN con sólo tres columnas: 
#   - Lexema
#   - Definición
#   - Información extra
diccionarioPLN = pd.DataFrame({
      'lexema': diccionario['Lexema'],
      'definición': diccionario['Definición'],
      'extra': diccionario[['Acepciones o nombre original',
                                  'Campo semántico',
                                  'Campo semántico 2', 
                                  'Complementación de la definición',
                                  'Situación histórica antes y durante la Ref.',
                                  'Relación-aportes a la Reforma']]\
                                  .apply(' '.join, axis=1)
                              })
list(diccionarioPLN.columns)

['lexema', 'definición', 'extra']

## PLN

Procesamiento de Lenguaje Natural, o *Natural Language Processing* (NLP)

In [9]:
# Usar modelo PLN en español
nlp = spacy.load('es_core_news_lg')

In [10]:
# Leer diccionario con modelo 
for i in diccionarioPLN:
  diccionarioPLN[i] = diccionarioPLN[i].apply(lambda x: nlp(x))

In [11]:
# Visualizar diccionario ya leído
diccionarioPLN.head()

lexema  \
0  (A, la, nobleza, cristiana, de, la, nación, al...   
1                                      (Agustino, .)   
2                          (Anabaptista, ,, iglesia)   
3                            (Anglicana, ,, Iglesia)   
4                                     (Arminiano, .)   

                                          definición  \
0  (Manifiesto, de, Lutero, contra, la, ICAR, ,, ...   
1   (Miembro, de, la, Orden, de,  , San, Agustín, .)   
2  (Movimiento, reformado, que, pugnó, por, un, b...   
3  (Movimiento, reformado, de, Inglaterra, con, m...   
4  (Movimiento, de, la, reforma, iniciado, por, A...   

                                               extra  
0  (An, den, christlichen, Adel, deutscher, Natio...  
1                       ( , Órdenes, católicas,    )  
2  (Anabautista, ;, Iglesia, anabaptista, Interpr...  
3  (Episcopal, ,, Iglesia, ;, Iglesia, anglicana,...  
4  (Wesleyano, ;, Interpretación, bíblica,  , La,...

Las palabras, separadas por comas, están encerradas entre paréntesis. Significa que el `diccionarioPLN` ya fue leído por el modelo en español

## Función para acceder al diccionario

In [12]:
# Ver lexemas con un índice
def verlexemas():
  data_table.enable_dataframe_formatter()
  return data_table.DataTable(diccionarioPLN[['lexema']], include_index=True, num_rows_per_page=20)
  data_table.disable_dataframe_formatter()

In [13]:
verlexemas()

In [14]:
# FUNCIÓN
# Consulta: 1 = 'lexema', 2 = 'definición'; 3 ='extra'
# Columna: Número de columna en index

def diccionario (consulta, columna):
  if consulta == 1:
    consulta = 'lexema'
  elif consulta == 2:
    consulta = 'definición'
  elif consulta == 3:
    consulta = 'extra'
  return diccionarioPLN[consulta][columna]

In [15]:
# Primer lexema
diccionario(consulta = 1, columna = 0)

A la nobleza cristiana de la nación alemana.

In [16]:
# Definición del primer lexema
diccionario(2,0)

Manifiesto de Lutero contra la ICAR, donde invita a Carlos V y a los principales de Alemania a reformar la Iglesia. |

In [17]:
# Información extra del primer lexema
diccionario(3,0)

An den christlichen Adel deutscher Nation von des christlichen Standes Besserung; Literatura reformada  Señala los tres muros del romanismo a derribar: supremacía del clero, monopolio interpretativo de las Escrituras, y supremacía del Papa para convocar concilios. Critica el lujo excesivo romano. Denuncia los métodos para el abuso financiero, arrendatario y también al Derecho Canónico. |  Lutero da 28 propuestas y a lo largo de ellas y en todo el escrito, presenta una pre-separación de iglesia y estado y una pre-desamortización de bienes eclesiásticos en Alemania. También frenar los impuestos y poderes a Roma, reducir asalariados , los lujos en Roma, juramentos, ceremonias, liturgias, regular las peregrinaciones, los conventos, votos y celibato, sacerdotes, misas, penas y el Derecho Canónico, fiestas y domingos, vínculos de parentesco, canonización, ventas de documentos eclesiales, mendicidad, relación con los bohemios, la didáctica y literatura en universidades, gobierno alemán, prést

In [18]:
diccionario(1,10)

Castillo fuerte es nuestro Dios.

## Ver vectorización

De acuerdo con la [documentación](https://spacy.io/models/es) en spaCy, estos vectores se forman en un proceso semántico que implicó 300 dimensiones.

In [19]:
diccionario(1,10).vector

array([-0.41945648,  0.6943    , -2.6311617 ,  0.17101835, -0.19501166,
       -2.2149951 ,  0.0346    ,  2.9261951 ,  0.22321832, -1.3064076 ,
       -0.63523996,  0.62290335,  0.06053168,  0.46940494, -0.45168665,
       -0.51289004,  1.6176237 ,  2.4277334 ,  0.13695832,  1.2706839 ,
       -0.08933838, -0.21290831, -0.919995  ,  0.31322166, -1.4326111 ,
       -1.0549166 ,  3.0710833 , -2.0453682 , -0.82245827, -0.569535  ,
        0.7332502 ,  1.068265  ,  0.8708187 ,  1.5303365 , -1.6829637 ,
       -0.77058953, -0.07880839,  2.4785936 , -1.2709919 , -1.1106117 ,
        0.16749   , -0.45886168, -0.272305  , -0.63645035, -2.1675549 ,
       -0.3997883 , -0.25131497, -3.2358    ,  0.34230837,  0.7768667 ,
        2.1007168 , -1.0968562 , -1.1701568 ,  0.3977584 ,  0.59559935,
        1.0687784 ,  0.34306172, -1.2061616 , -1.6674519 , -0.2451588 ,
       -0.55529207, -0.6286667 ,  0.16287005, -0.64427835, -0.42882332,
        1.7884833 , -2.1311448 , -0.03818838,  0.45687422,  0.01

# Pregunta natural

In [20]:
# Función para recibir pregunta del usuario
def pregunta():
  global preguntaPLN
  preguntaPLN = nlp(input('''Si no recuerdas el término, pregúntame, o trata de escribir lo que recuerdes:
(Nota: El diccionario cuenta con 70 términos. Trabajamos para que crezca)
→ '''))

In [21]:
# Preguntar: Documento que quemó Martín Lutero
pregunta()

Si no recuerdas el término, pregúntame, o trata de escribir lo que recuerdes:
(Nota: El diccionario cuenta con 70 términos. Trabajamos para que crezca)
→ Documento que quemó Martín Lutero


La respuesta queda guardada como `preguntaPLN`

# Respuesta al usuario

## Vectores
Cuando se vectorizan las palabras pueden compararse sus listas de vectores para encontrar la mayor coincidencia semántica, donde semántica se refiere a que aparecen en los mismos contextos.
En este modelo se utilizará el módulo [`es_core_news_lg`](https://spacy.io/models/es). Es un modelo muy preciso; la desventaja es su tamaño en comparación con `es_core_news_md`, que es mucho más pequeño.

In [22]:
# Ver similitud por vectores
print(f'''Comparación
"Martín Lutero lee la Biblia" vs "Martín Lutero lee las Sagradas Escrituras"
Porcentaje de semejanza
{round(nlp("Martín Lutero lee la Biblia").similarity(nlp("Anita lava la bañera"))*100, 2)} %
Comparación
"Martín Lutero lee la Biblia" vs "Martín Lutero lee las 95 tesis"
Porcentaje de semejanza
{round(nlp("Martín Lutero lee la Biblia").similarity(nlp("Martín Lutero lee las 95 tesis"))*100, 2)} %
Comparación
"Martín Lutero lee la Biblia" vs "Martín Lutero escribe un himno"
Porcentaje de semejanza
{round(nlp("Martín Lutero lee la Biblia").similarity(nlp("Martín Lutero escribe un himno"))*100, 2)} %
Comparación
"Martín Lutero lee la Biblia" vs "Martín Lutero es un hereje"
Porcentaje de semejanza
{round(nlp("Martín Lutero lee la Biblia").similarity(nlp("Martín Lutero es un hereje"))*100, 2)} %''')

Comparación
"Martín Lutero lee la Biblia" vs "Martín Lutero lee las Sagradas Escrituras"
Porcentaje de semejanza
59.89 %
Comparación
"Martín Lutero lee la Biblia" vs "Martín Lutero lee las 95 tesis"
Porcentaje de semejanza
42.63 %
Comparación
"Martín Lutero lee la Biblia" vs "Martín Lutero escribe un himno"
Porcentaje de semejanza
26.19 %
Comparación
"Martín Lutero lee la Biblia" vs "Martín Lutero es un hereje"
Porcentaje de semejanza
20.94 %


## Función para comparar preguntaPLN vs diccionarioPLN

In [43]:
# Función para comparar pregunta con el diccionario
def compara():
  global diccionarioPLN
  global preguntaPLN
  for i in diccionarioPLN: 
    diccionarioPLN['similitudLexema'] = diccionarioPLN['lexema'].apply(lambda x: x.similarity(preguntaPLN))
  for i in diccionarioPLN:
    diccionarioPLN['similitudDefinición'] = diccionarioPLN['definición'].apply(lambda x: x.similarity(preguntaPLN))
  for i in diccionarioPLN:
    diccionarioPLN['similitudExtra'] = diccionarioPLN['extra'].apply(lambda x: x.similarity(preguntaPLN))

In [44]:
# Comparar
compara()
# Ver resultados
diccionarioPLN.head()

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:8: UserWarning: [W008] Evaluating Doc.similarity based on empty vectors.
  
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:10: UserWarning: [W008] Evaluating Doc.similarity based on empty vectors.
  # Remove the CWD from sys.path while we load stuff.


lexema  \
0  (A, la, nobleza, cristiana, de, la, nación, al...   
1                                      (Agustino, .)   
2                          (Anabaptista, ,, iglesia)   
3                            (Anglicana, ,, Iglesia)   
4                                     (Arminiano, .)   

                                          definición  \
0  (Manifiesto, de, Lutero, contra, la, ICAR, ,, ...   
1   (Miembro, de, la, Orden, de,  , San, Agustín, .)   
2  (Movimiento, reformado, que, pugnó, por, un, b...   
3  (Movimiento, reformado, de, Inglaterra, con, m...   
4  (Movimiento, de, la, reforma, iniciado, por, A...   

                                               extra  similitudDefinición  \
0  (An, den, christlichen, Adel, deutscher, Natio...             0.296707   
1                       ( , Órdenes, católicas,    )             0.189481   
2  (Anabautista, ;, Iglesia, anabaptista, Interpr...             0.496899   
3  (Episcopal, ,, Iglesia, ;, Iglesia, anglicana,...             0.286716   
4  (Wesleyano, ;, Interpretación, bíblica,  , La,...             0.342929   

   similitudExtra  similitudLexema  
0        0.407636         0.227601  
1        0.360645         0.008994  
2        0.388825         0.458843  
3        0.477696         0.476516  
4        0.373309        -0.171416

## Función para mostrar resultados al usuario

In [45]:
def resultados():
  lexema = diccionarioPLN.sort_values(by='similitudLexema', ascending=False).reset_index().drop(columns='index')
  definir = diccionarioPLN.sort_values(by='similitudDefinición', ascending=False).reset_index().drop(columns='index')
  extra = diccionarioPLN.sort_values(by='similitudExtra', ascending=False).reset_index().drop(columns='index')

  # Lexema
  print('\nBúsqueda en lista de lexemas. 10 coincidencias más cercanas:\n')
  for l in range(0,10):
    l += 0
    print(f'''{l+1}. {lexema.loc[l, 'lexema']}:   {round(lexema.loc[l, 'similitudLexema']*100, 2)} % de similitud''')

  # Definiciones
  print('\nBúsqueda en lista de definiciones. 10 coincidencias más cercanas:\n')
  for d in range(0,9):
    d += 0
    print(f'''{d+1}. {definir.loc[d, 'lexema']}:  {round(definir.loc[d, 'similitudDefinición']*100, 2)} % de similitud''')
  
  # Extra
  print('\n\nBúsqueda en lista de información extra. 10 coincidencias más cercanas:\n')
  for e in range(0,9):
    e += 0
    print(f'''{e+1}. {extra.loc[e, 'lexema']}:   {round(extra.loc[e, 'similitudExtra']*100, 2)} % de similitud''')
  print('\n')

In [46]:
resultados()


Búsqueda en lista de lexemas. 10 coincidencias más cercanas:

1. Sacramento:   48.38 % de similitud
2. Anglicana, Iglesia:   47.65 % de similitud
3. Anabaptista, iglesia:   45.88 % de similitud
4. A la nobleza cristiana de la nación alemana.:   22.76 % de similitud
5. Cautividad babilónica de la Iglesia, La:   21.77 % de similitud
6. Institución de la Religión Cristiana.:   20.61 % de similitud
7. De la libertad del cristiano.:   19.28 % de similitud
8. Valdense, grupo.:   19.24 % de similitud
9. Reforma Protestante.:   18.73 % de similitud
10. Protesta de Espira:   18.62 % de similitud

Búsqueda en lista de definiciones. 10 coincidencias más cercanas:

1. Cinco solas. Sola fide.:  55.46 % de similitud
2. Valdense, grupo.:  52.28 % de similitud
3. Anabaptista, iglesia:  49.69 % de similitud
4. Husita, grupo.:  47.79 % de similitud
5. Lolarda, grupo.:  47.3 % de similitud
6. Exsurge Domine.:  47.09 % de similitud
7. Calvinismo.:  46.69 % de similitud
8. Libre examen.:  46.1 % de simili

El "Documento que quemó Martín Lutero" que yo tenía en mente era la bula *Exsurge Domine*:
 - En **lexemas** no aparece
 - En **definiciones** sí aparece 
  - 6o lugar
  - 47.09 % de similitud
 - En **información extra** sí aparece
  - 7o lugar
  - 46.22 % de similitud

In [47]:
verlexemas()

In [48]:
diccionario(1,42)

Exsurge Domine.

In [49]:
diccionario(3,42)

Bula contra los errores de Martín Lutero y secuaces (Bulla contra errores Martini Lutheri [et] sequacium), Literatura católica  Redactada con mucho cuidado  y escrutinio por León X con Juan Eck, Benedetto Accolti, Julio de Médici (éste último sería Clemente VII) y la curia romana, esgrimía 41 puntos erráticos de Lutero sobre las indulgencias, jerarquía eclesiástica, pecado, penitencia, purgatorio, sacramentos, entre otros. El texto comienza: “Levántate, oh Dios, y juzga tu propia causa”, aludiendo a las bulas de excomunión anteriores. Apela a Dios, luego a Pedro, Pablo, la Iglesia y, en última instancia, a la nación alemana. | Al momento de su publicación, era muy evidente la separación de Lutero. La bula atestigua una citación anterior que Lutero rechazó, nominándolo “contumaz” y comunica su negativa a creer en los concilios. A pesar de esto, se le anima a regresar a la iglesia y se le promete una cálida bienvenida en su calidad de hijo pródigo. | Lutero confirma y comunica sus decisi

# Preguntas de evaluación

Se harán algunas preguntas para ver el comportamiento del modelo.

El lexema, la definición y la información extra arrojan resultados diferentes porque contienen información diferente.

In [50]:
# Función pregunta-respuesta
def diccionarioresponde():
  pregunta()
  compara()
  resultados()


## Preguntas con contenido ajeno al diccionario

### La pizza más rica del mundo

In [51]:
# La pizza más rica del mundo
diccionarioresponde()

Si no recuerdas el término, pregúntame, o trata de escribir lo que recuerdes:
(Nota: El diccionario cuenta con 70 términos. Trabajamos para que crezca)
→ La pizza más rica del mundo

Búsqueda en lista de lexemas. 10 coincidencias más cercanas:

1. Cautividad babilónica de la Iglesia, La:   75.26 % de similitud
2. A la nobleza cristiana de la nación alemana.:   65.19 % de similitud
3. Libro de la Concordia.:   62.12 % de similitud
4. Institución de la Religión Cristiana.:   60.84 % de similitud
5. De la libertad del cristiano.:   57.43 % de similitud
6. Salterio de Ginebra.:   44.73 % de similitud
7. Protesta de Espira:   44.41 % de similitud
8. Libro de oración común.:   44.35 % de similitud
9. Disputa de Leipzig.:   41.9 % de similitud
10. Dieta. Dieta de Augsburgo.:   40.64 % de similitud

Búsqueda en lista de definiciones. 10 coincidencias más cercanas:

1. Eucaristía. Consubstanciación.:  63.07 % de similitud
2. Consenso Tigurinus.:  59.7 % de similitud
3. Cinco solas. Soli Deo Glo

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:8: UserWarning: [W008] Evaluating Doc.similarity based on empty vectors.
  
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:10: UserWarning: [W008] Evaluating Doc.similarity based on empty vectors.
  # Remove the CWD from sys.path while we load stuff.


### Qué marca de auto es "Mi rebelde"?

In [52]:
# Qué marca de auto es "Mi rebelde"?
diccionarioresponde()

Si no recuerdas el término, pregúntame, o trata de escribir lo que recuerdes:
(Nota: El diccionario cuenta con 70 términos. Trabajamos para que crezca)
→ Qué marca de auto es "Mi rebelde"?

Búsqueda en lista de lexemas. 10 coincidencias más cercanas:

1. Castillo fuerte es nuestro Dios.:   50.72 % de similitud
2. Exsurge Domine.:   35.63 % de similitud
3. De la libertad del cristiano.:   34.72 % de similitud
4. Canon. Canon bíblico.:   31.98 % de similitud
5. Barba.:   31.24 % de similitud
6. Calvinismo.:   30.8 % de similitud
7. Dieta. Dieta de Worms.:   29.97 % de similitud
8. Canon.:   29.52 % de similitud
9. Eucaristía. Anamnesis.:   28.81 % de similitud
10. Dieta. Dieta de Augsburgo.:   28.42 % de similitud

Búsqueda en lista de definiciones. 10 coincidencias más cercanas:

1. Cinco solas. Sola gratia.:  33.74 % de similitud
2. Cinco solas. Solus Christus.:  31.38 % de similitud
3. Cinco puntos calvinistas. 4. Irresistible gracia.:  29.27 % de similitud
4. Cinco puntos calvinistas

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:8: UserWarning: [W008] Evaluating Doc.similarity based on empty vectors.
  
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:10: UserWarning: [W008] Evaluating Doc.similarity based on empty vectors.
  # Remove the CWD from sys.path while we load stuff.


## Preguntas con contenido relacionado, pero sin el lexema al que se alude

### Biblia traducida por los Países Bajos

In [53]:
# Biblia traducida por los Países Bajos
# Se busca "Statenvertaling", que aún no está en el diccionario
diccionarioresponde()

Si no recuerdas el término, pregúntame, o trata de escribir lo que recuerdes:
(Nota: El diccionario cuenta con 70 términos. Trabajamos para que crezca)
→ Biblia traducida por los Países Bajos

Búsqueda en lista de lexemas. 10 coincidencias más cercanas:

1. Guerra de los ochenta años.:   54.19 % de similitud
2. Anglicana, Iglesia:   49.52 % de similitud
3. Anabaptista, iglesia:   43.24 % de similitud
4. Noventa y cinco tesis.:   42.79 % de similitud
5. Cinco puntos calvinistas. 5. Perseverancia de los Santos.:   41.97 % de similitud
6. Cinco puntos calvinistas.:   40.95 % de similitud
7. Libro de oración común.:   40.45 % de similitud
8. Edicto de Nantes.:   38.38 % de similitud
9. Institución de la Religión Cristiana.:   34.98 % de similitud
10. Salterio de Ginebra.:   34.67 % de similitud

Búsqueda en lista de definiciones. 10 coincidencias más cercanas:

1. Dieta. Dieta de Worms.:  65.42 % de similitud
2. Libre examen.:  62.71 % de similitud
3. Guerra de los ochenta años.:  61.9 % d

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:8: UserWarning: [W008] Evaluating Doc.similarity based on empty vectors.
  
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:10: UserWarning: [W008] Evaluating Doc.similarity based on empty vectors.
  # Remove the CWD from sys.path while we load stuff.


### El padre de la reforma protestante

In [54]:
# El padre de la reforma protestante
# Se busca "Martín Lutero", que aún no está en el diccionario
diccionarioresponde()

Si no recuerdas el término, pregúntame, o trata de escribir lo que recuerdes:
(Nota: El diccionario cuenta con 70 términos. Trabajamos para que crezca)
→ El padre de la reforma protestante

Búsqueda en lista de lexemas. 10 coincidencias más cercanas:

1. Cautividad babilónica de la Iglesia, La:   75.82 % de similitud
2. De la libertad del cristiano.:   73.76 % de similitud
3. Institución de la Religión Cristiana.:   71.92 % de similitud
4. Libro de la Concordia.:   70.02 % de similitud
5. A la nobleza cristiana de la nación alemana.:   69.35 % de similitud
6. Protesta de Espira:   67.23 % de similitud
7. Salterio de Ginebra.:   60.77 % de similitud
8. Guerra de Esmalcalda.:   57.69 % de similitud
9. Sínodo de Dort.:   57.37 % de similitud
10. Paz de Augsburgo.:   57.13 % de similitud

Búsqueda en lista de definiciones. 10 coincidencias más cercanas:

1. Luteranismo.:  82.01 % de similitud
2. Castillo fuerte es nuestro Dios.:  76.3 % de similitud
3. Eucaristía. Consubstanciación.:  76.2

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:8: UserWarning: [W008] Evaluating Doc.similarity based on empty vectors.
  
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:10: UserWarning: [W008] Evaluating Doc.similarity based on empty vectors.
  # Remove the CWD from sys.path while we load stuff.


## Preguntas con el lexema aludido dentro del diccionario

### Batallas entre católicos y protestantes

In [55]:
# Batallas entre católicos y protestantes
diccionarioresponde()

Si no recuerdas el término, pregúntame, o trata de escribir lo que recuerdes:
(Nota: El diccionario cuenta con 70 términos. Trabajamos para que crezca)
→ Batallas entre católicos y protestantes

Búsqueda en lista de lexemas. 10 coincidencias más cercanas:

1. Noventa y cinco tesis.:   66.34 % de similitud
2. Anglicana, Iglesia:   64.24 % de similitud
3. Guerra de los ochenta años.:   64.16 % de similitud
4. Cinco puntos calvinistas.:   63.01 % de similitud
5. Anabaptista, iglesia:   60.29 % de similitud
6. Cinco puntos calvinistas. 5. Perseverancia de los Santos.:   59.2 % de similitud
7. Edicto de Nantes.:   58.97 % de similitud
8. Lolarda, grupo.:   57.61 % de similitud
9. Huterita, grupo.:   57.61 % de similitud
10. Husita, grupo.:   57.61 % de similitud

Búsqueda en lista de definiciones. 10 coincidencias más cercanas:

1. Decet romanum pontificem.:  78.53 % de similitud
2. Calvinismo.:  76.03 % de similitud
3. Libre examen.:  75.55 % de similitud
4. Dieta. Dieta de Worms.:  74.65 

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:8: UserWarning: [W008] Evaluating Doc.similarity based on empty vectors.
  
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:10: UserWarning: [W008] Evaluating Doc.similarity based on empty vectors.
  # Remove the CWD from sys.path while we load stuff.


In [56]:
# ¿Dónde se enfrentó Lutero a Carlos quinto?
diccionarioresponde()

Si no recuerdas el término, pregúntame, o trata de escribir lo que recuerdes:
(Nota: El diccionario cuenta con 70 términos. Trabajamos para que crezca)
→ ¿Dónde se enfrentó Lutero a Carlos quinto?

Búsqueda en lista de lexemas. 10 coincidencias más cercanas:

1. A la nobleza cristiana de la nación alemana.:   18.62 % de similitud
2. Sacramento:   15.03 % de similitud
3. Anglicana, Iglesia:   14.41 % de similitud
4. Calvinismo.:   13.72 % de similitud
5. Luteranismo.:   13.51 % de similitud
6. De la libertad del cristiano.:   12.97 % de similitud
7. Anabaptista, iglesia:   12.49 % de similitud
8. Cisma.:   12.4 % de similitud
9. Reforma Protestante.:   10.54 % de similitud
10. Noventa y cinco tesis.:   10.42 % de similitud

Búsqueda en lista de definiciones. 10 coincidencias más cercanas:

1. Valdense, grupo.:  37.48 % de similitud
2. Cinco puntos calvinistas. 1. Totalidad depravada.:  37.02 % de similitud
3. Exsurge Domine.:  36.01 % de similitud
4. Husita, grupo.:  33.4 % de similitud

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:8: UserWarning: [W008] Evaluating Doc.similarity based on empty vectors.
  
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:10: UserWarning: [W008] Evaluating Doc.similarity based on empty vectors.
  # Remove the CWD from sys.path while we load stuff.


## ¿Siempre aparecere en los primeros 5 resultados?

### Escrito de martín lutero a los principales alemanes

In [57]:
# Escrito de martín lutero a los principales alemanes
diccionarioresponde()

Si no recuerdas el término, pregúntame, o trata de escribir lo que recuerdes:
(Nota: El diccionario cuenta con 70 términos. Trabajamos para que crezca)
→ Escrito de martín lutero a los principales alemanes

Búsqueda en lista de lexemas. 10 coincidencias más cercanas:

1. Guerra de los ochenta años.:   67.86 % de similitud
2. Edicto de Nantes.:   59.54 % de similitud
3. Protesta de Espira:   59.04 % de similitud
4. A la nobleza cristiana de la nación alemana.:   57.71 % de similitud
5. Libro de oración común.:   56.99 % de similitud
6. Cinco puntos calvinistas. 5. Perseverancia de los Santos.:   56.69 % de similitud
7. Salterio de Ginebra.:   55.9 % de similitud
8. Guerra de Esmalcalda.:   55.72 % de similitud
9. Institución de la Religión Cristiana.:   54.88 % de similitud
10. Sínodo de Dort.:   53.98 % de similitud

Búsqueda en lista de definiciones. 10 coincidencias más cercanas:

1. A la nobleza cristiana de la nación alemana.:  84.38 % de similitud
2. Matanza de San Bartolomé.:  81

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:8: UserWarning: [W008] Evaluating Doc.similarity based on empty vectors.
  
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:10: UserWarning: [W008] Evaluating Doc.similarity based on empty vectors.
  # Remove the CWD from sys.path while we load stuff.


### No: Precursores de la reforma

In [58]:
# Precursores de la reforma
diccionarioresponde()

Si no recuerdas el término, pregúntame, o trata de escribir lo que recuerdes:
(Nota: El diccionario cuenta con 70 términos. Trabajamos para que crezca)
→ Precursores de la reforma

Búsqueda en lista de lexemas. 10 coincidencias más cercanas:

1. Institución de la Religión Cristiana.:   90.23 % de similitud
2. Cautividad babilónica de la Iglesia, La:   88.03 % de similitud
3. A la nobleza cristiana de la nación alemana.:   87.72 % de similitud
4. Libro de la Concordia.:   87.29 % de similitud
5. Protesta de Espira:   75.63 % de similitud
6. De la libertad del cristiano.:   71.51 % de similitud
7. Salterio de Ginebra.:   67.06 % de similitud
8. Libro de oración común.:   66.75 % de similitud
9. Sínodo de Dort.:   64.34 % de similitud
10. Edicto de Nantes.:   63.82 % de similitud

Búsqueda en lista de definiciones. 10 coincidencias más cercanas:

1. Eucaristía. Consubstanciación.:  86.77 % de similitud
2. Protesta de Espira:  85.74 % de similitud
3. Libro de la Concordia.:  85.59 % de sim

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:8: UserWarning: [W008] Evaluating Doc.similarity based on empty vectors.
  
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:10: UserWarning: [W008] Evaluating Doc.similarity based on empty vectors.
  # Remove the CWD from sys.path while we load stuff.


## ¿Se puede prescindir la lista de lexemas del análisis?

In [59]:
verlexemas()

### Tipo de monje que era martín Lutero

In [60]:
# Tipo de monje que era martín Lutero
diccionarioresponde()

Si no recuerdas el término, pregúntame, o trata de escribir lo que recuerdes:
(Nota: El diccionario cuenta con 70 términos. Trabajamos para que crezca)
→ Tipo de monje que era martín Lutero

Búsqueda en lista de lexemas. 10 coincidencias más cercanas:

1. Libro de oración común.:   56.3 % de similitud
2. Salterio de Ginebra.:   56.18 % de similitud
3. Concilio. Concilio de Trento.:   55.2 % de similitud
4. Sínodo de Dort.:   54.62 % de similitud
5. Edicto de Nantes.:   53.69 % de similitud
6. Guerra de Esmalcalda.:   53.32 % de similitud
7. Protesta de Espira:   52.56 % de similitud
8. Institución de la Religión Cristiana.:   52.13 % de similitud
9. Anabaptista, iglesia:   52.04 % de similitud
10. A la nobleza cristiana de la nación alemana.:   51.98 % de similitud

Búsqueda en lista de definiciones. 10 coincidencias más cercanas:

1. Institución de la Religión Cristiana.:  65.1 % de similitud
2. Paz de Augsburgo.:  64.6 % de similitud
3. Luteranismo.:  64.2 % de similitud
4. Libro de 

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:8: UserWarning: [W008] Evaluating Doc.similarity based on empty vectors.
  
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:10: UserWarning: [W008] Evaluating Doc.similarity based on empty vectors.
  # Remove the CWD from sys.path while we load stuff.


In [61]:
diccionario(1,1)

Agustino.

In [62]:
diccionario(2,1)

Miembro de la Orden de  San Agustín.

In [63]:
diccionario(1,58)

Orden de San Agustín.

In [64]:
diccionario(2,58)

Orden regida por la Regla de San Agustín, formada de varias comunidades de monjes. |

In [65]:
diccionario(3,58)

Orden de Ermitaños de San Agustín; Orden de los Agustinos    Eran comunitarios, austeros, rígidos en salud, castidad y obediencia. | Estaban divididos en 4 provincias: Italia, España, Francia y Alemania. Para la segunda mitad del siglo XV la orden estaba dividida en dos: Los que querían una reforma en la orden, llamados no observantes,  y los que se apegaban a su orden estricto, los observantes. La congregación de Erfurt se inclinó por la observancia entricta desde 1492. | Martín Lutero pertenecía a ésta última. Entre 1510 y 1511 fue enviado con otro monje a Roma para representar a todos los monasterios agustinos de Alemania para defender su posición ante el papa Julio II, quien finalmente fusionó las órdenes observantes y no observantes. Es en este periodo cuando la moralidad y espiritualidad en Roma le causan una impresión negativa |

### Documento donde se excomulga a Lutero

In [66]:
# Documento donde se excomulga a Lutero
diccionarioresponde()

Si no recuerdas el término, pregúntame, o trata de escribir lo que recuerdes:
(Nota: El diccionario cuenta con 70 términos. Trabajamos para que crezca)
→ Documento donde se excomulga a Lutero

Búsqueda en lista de lexemas. 10 coincidencias más cercanas:

1. Sacramento:   41.91 % de similitud
2. Anglicana, Iglesia:   37.25 % de similitud
3. Anabaptista, iglesia:   34.28 % de similitud
4. A la nobleza cristiana de la nación alemana.:   33.88 % de similitud
5. Cautividad babilónica de la Iglesia, La:   22.1 % de similitud
6. Institución de la Religión Cristiana.:   21.0 % de similitud
7. Noventa y cinco tesis.:   18.26 % de similitud
8. Guerra de los ochenta años.:   18.07 % de similitud
9. Valdense, grupo.:   15.93 % de similitud
10. De la libertad del cristiano.:   14.99 % de similitud

Búsqueda en lista de definiciones. 10 coincidencias más cercanas:

1. Valdense, grupo.:  69.61 % de similitud
2. Husita, grupo.:  60.68 % de similitud
3. Eucaristía.:  59.12 % de similitud
4. Exsurge Dom

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:8: UserWarning: [W008] Evaluating Doc.similarity based on empty vectors.
  
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:10: UserWarning: [W008] Evaluating Doc.similarity based on empty vectors.
  # Remove the CWD from sys.path while we load stuff.


### Escrito más famoso de Martín Lutero

In [67]:
# Escrito más famoso de Martín Lutero
diccionarioresponde()

Si no recuerdas el término, pregúntame, o trata de escribir lo que recuerdes:
(Nota: El diccionario cuenta con 70 términos. Trabajamos para que crezca)
→ Escrito más famoso de Martín Lutero

Búsqueda en lista de lexemas. 10 coincidencias más cercanas:

1. Protesta de Espira:   65.98 % de similitud
2. Salterio de Ginebra.:   63.38 % de similitud
3. Libro de oración común.:   62.09 % de similitud
4. Guerra de Esmalcalda.:   60.45 % de similitud
5. Edicto de Nantes.:   59.99 % de similitud
6. Paz de Augsburgo.:   59.03 % de similitud
7. Sínodo de Dort.:   58.57 % de similitud
8. Disputa de Leipzig.:   58.51 % de similitud
9. Guerra de los ochenta años.:   58.03 % de similitud
10. Matanza de San Bartolomé.:   58.01 % de similitud

Búsqueda en lista de definiciones. 10 coincidencias más cercanas:

1. Anglicana, Iglesia:  67.15 % de similitud
2. Noventa y cinco tesis.:  66.02 % de similitud
3. Calvinismo.:  65.41 % de similitud
4. Castillo fuerte es nuestro Dios.:  64.83 % de similitud
5. Li

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:8: UserWarning: [W008] Evaluating Doc.similarity based on empty vectors.
  
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:10: UserWarning: [W008] Evaluating Doc.similarity based on empty vectors.
  # Remove the CWD from sys.path while we load stuff.


### Sí, a falta de mayor análisis.

Faltaría un mayor análisis para confirmarlo, pero por el momento sí.

# Reduciendo función

- Se reduce un 33 % los resultados
- Es más sencillo leer 20 resultados a 30
- Los resultados ignorados es muy problabe que no sean necesarios

In [68]:
def resultados2():
  lexema = diccionarioPLN.sort_values(by='similitudLexema', ascending=False).reset_index().drop(columns='index')
  definir = diccionarioPLN.sort_values(by='similitudDefinición', ascending=False).reset_index().drop(columns='index')
  extra = diccionarioPLN.sort_values(by='similitudExtra', ascending=False).reset_index().drop(columns='index')

  # Definiciones
  print('\nBúsqueda en lista de definiciones. 10 coincidencias más cercanas:\n')
  for d in range(0,9):
    d += 0
    print(f'''{d+1}. {definir.loc[d, 'lexema']}:  {round(definir.loc[d, 'similitudDefinición']*100, 2)} % de similitud''')
  
  # Extra
  print('\n\nBúsqueda en lista de información extra. 10 coincidencias más cercanas:\n')
  for e in range(0,9):
    e += 0
    print(f'''{e+1}. {extra.loc[e, 'lexema']}:   {round(extra.loc[e, 'similitudExtra']*100, 2)} % de similitud''')
  print('\n')

## Función final: diccionarioContesta()

In [69]:
# Función pregunta-respuesta 2
def diccionarioContesta():
  pregunta()
  compara()
  resultados2()

In [70]:
# Cómo se llama cuando el pan y el vino se hacen el cuerpo de Cristo
diccionarioContesta()

Si no recuerdas el término, pregúntame, o trata de escribir lo que recuerdes:
(Nota: El diccionario cuenta con 70 términos. Trabajamos para que crezca)
→ Cómo se llama cuando el pan y el vino se hacen el cuerpo de Cristo

Búsqueda en lista de definiciones. 10 coincidencias más cercanas:

1. Sacramento:  72.87 % de similitud
2. Paz de Augsburgo.:  66.4 % de similitud
3. Cinco solas. Sola fide.:  65.87 % de similitud
4. Eucaristía. Transubstanciación.:  65.64 % de similitud
5. Anabaptista, iglesia:  64.11 % de similitud
6. Eucaristía.:  62.68 % de similitud
7. Indulgencia.:  62.55 % de similitud
8. Cinco puntos calvinistas. 1. Totalidad depravada.:  60.79 % de similitud
9. Exsurge Domine.:  60.3 % de similitud


Búsqueda en lista de información extra. 10 coincidencias más cercanas:

1. Cautividad babilónica de la Iglesia, La:   75.99 % de similitud
2. Coral.:   73.1 % de similitud
3. Eucaristía.:   67.84 % de similitud
4. Disputa de Leipzig.:   65.83 % de similitud
5. Husita, grupo.:   6

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:8: UserWarning: [W008] Evaluating Doc.similarity based on empty vectors.
  
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:10: UserWarning: [W008] Evaluating Doc.similarity based on empty vectors.
  # Remove the CWD from sys.path while we load stuff.


# Conclusión

|**Búsquedas exitosas:**|**Búsquedas fallidas**|
|---|---|
**Documento que quemó Martín Lutero** | **Tipo de monje que era martín Lutero**
*- "Exsurge Domine"* |*- Se esperaba "Orden de San Agustín"*
**Batallas entre católicos y protestantes**
*- "Guerra de los ochenta años". No mostró "Guerra de Esmalcalda"*
**¿Dónde se enfrentó Lutero a Carlos quinto?**
*- "Dieta de Worms"*
**Escrito de martín lutero a los principales alemanes**
*- "A la nobleza cristiana de la nación alemana"*
**Precursores de la reforma (parcial)**
*- "Husita", "Remonstrante", "Gomarista". Faltó "Valdense"*
**Documento donde se excomulga a Lutero**
*- "Decet Romanum Pontificem"*
**Escrito más famoso de Martín Lutero**
*- "95 tesis"*
**- Cómo se llama cuando el pan y el vino se hacen el cuerpo de Cristo**
*- "Eucaristía. Transubstanciación"*